### PR TITLE
WIP: World builder 2.0

### PR DESCRIPTION
**Pull request by MFraters**
This replaces #2122. It is now a semi-separate library, which doesn't use any external functions except boost. The license is set a lgplv3, so that is possible for other codes to use it without changing their licence. They do have to share any modifications to the library which they made. Is anyone sees any copyright concerns in the files, please let me know.

I have removed the support for ellipsoidal for now, because I would need to contact Sarah Stamps first for permission. The only copied code is I think the cartesian to spherical and back, but as I understood from @tjhei, it is too small and trivial to be copyrighted. Is this correct, or should I also rewrite this?

The astyle doesn't work yet, so I will take a look at that, and I will also add tests and try to add unit tests.

If there are issues of any other kind, please let me know.

And ofcourse some figures of what this can do (temperature plots):
screenshot from 2018-06-25 17-52-28
screenshot from 2018-06-25 17-52-44

### CODE REVIEW

File: contrib/world_builder/include/world_builder/coordinate_systems/interface.h [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+      public:
+        /**
+         * constructor
+         */
+        Interface();
+
+        /**
+         * Destructor
+         */
+        ~Interface();
```
Comment (tjhei) [inline contrib/world_builder/include/world_builder/coordinate_systems/interface.h]: all interface classes in c++ should have a virtual destructor. https://stackoverflow.com/a/461224

File: contrib/world_builder/include/world_builder/features/continental_plate.h [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+#include <world_builder/features/interface.h>
+#include <world_builder/world.h>
+
+
+namespace WorldBuilder
+{
+  namespace Features
+  {
+
+    class ContinentalPlate : public Interface
```
Comment (tjhei) [inline contrib/world_builder/include/world_builder/features/continental_plate.h]: document

File: contrib/world_builder/include/world_builder/features/continental_plate.h [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+         * Read in the world builder file
+         */
+        virtual
+        void read(ptree &property_tree);
+
+        /**
+         * Returns a temperature based on the given position
+         */
+        virtual
+        double temperature(const Point<3> position,
```
Comment (tjhei) [inline contrib/world_builder/include/world_builder/features/continental_plate.h]: pass point as a const &

File: contrib/world_builder/include/world_builder/features/continental_plate.h [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+                           const double depth,
+                           const double gravity,
+                           double temperature) const;
+
+        /**
+         * Returns a value for the reqeusted composition (0 is not present,
+         * 1 is present) based on the given position and
+         */
+        virtual
+        double composition(const Point<3> position,
```
Comment (tjhei) [inline contrib/world_builder/include/world_builder/features/continental_plate.h]: same here

File: contrib/world_builder/include/world_builder/features/continental_plate.h [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+         * Returns a temperature based on the given position
+         */
+        virtual
+        double temperature(const Point<3> position,
+                           const double depth,
+                           const double gravity,
+                           double temperature) const;
+
+        /**
+         * Returns a value for the reqeusted composition (0 is not present,
```
Comment (tjhei) [inline contrib/world_builder/include/world_builder/features/continental_plate.h]: typo "requested"

File: contrib/world_builder/include/world_builder/features/continental_plate.h [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+         * 1 is present) based on the given position and
+         */
+        virtual
```
Comment (tjhei) [inline contrib/world_builder/include/world_builder/features/continental_plate.h]: you could change this to return a bool instead and convert it to double inside the plugin. Maybe that is a more natural interface?

File: contrib/world_builder/include/world_builder/features/interface.h [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+
+
+namespace WorldBuilder
+{
+  class World;
+
+  namespace Features
+  {
+
+    class Interface
```
Comment (tjhei) [inline contrib/world_builder/include/world_builder/features/interface.h]: document

File: contrib/world_builder/include/world_builder/features/interface.h [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+      public:
+        /**
+         * constructor
+         */
+        Interface();
+
+        /**
+         * Destructor
+         */
+        ~Interface();
```
Comment (tjhei) [inline contrib/world_builder/include/world_builder/features/interface.h]: virtual

File: contrib/world_builder/include/world_builder/features/interface.h [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+
+        /**
+         * read in the world builder file
+         */
+        virtual
+        void read(ptree &property_tree) = 0;
```
Comment (tjhei) [inline contrib/world_builder/include/world_builder/features/interface.h]: can this be const? I assume you don't need to modify the propertytree?

File: contrib/world_builder/include/world_builder/features/interface.h [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+        std::string composition_submodule_name;
+
+    };
+
+
+    /**
+     * factory function
+     */
+    Interface *
+    create_feature(const std::string name, WorldBuilder::World *world);
```
Comment (tjhei) [inline contrib/world_builder/include/world_builder/features/interface.h]: const & name

Comment (tjhei) [reply contrib/world_builder/include/world_builder/features/interface.h]: can the world be NULL? if not, pass a reference.

File: contrib/world_builder/include/world_builder/point.h [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+
+#include <array>
+#include <cmath>
+
+#include <world_builder/coordinate_system.h>
+
+namespace WorldBuilder
+{
+  template<int dim>
+  class Point
```
Comment (tjhei) [inline contrib/world_builder/include/world_builder/point.h]: document

File: contrib/world_builder/include/world_builder/point.h [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+       * Constructor. Constructs a Point at (0,0) in 2d or (0,0,0) in 3d
+       * with a Cartesian coordinate system.
+       */
+      Point();
+
+      /**
+       * Constructor. Constructs a Point from a std::array<double,dim> and
+       * a coordinate system which is set by default to Cartesian.
+       */
+      Point(const std::array<double,dim> &array, CoordinateSystem coordinate_system = CoordinateSystem::cartesian);
```
Comment (tjhei) [inline contrib/world_builder/include/world_builder/point.h]: I find it confusing to name a variable "array"

File: contrib/world_builder/include/world_builder/point.h [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+       * a coordinate system which is set by default to Cartesian.
+       */
+      Point(const double x, const double y, const double z, CoordinateSystem coordinate_system = CoordinateSystem::cartesian);
+
+      /**
+       * Destructor
+       */
+      ~Point();
+
+      Point<dim> operator=(const Point<dim> &point);
```
Comment (tjhei) [inline contrib/world_builder/include/world_builder/point.h]: should return a Point<dim>&

File: contrib/world_builder/include/world_builder/point.h [hunk 14]
```diff
@@ -0,0 +1,167 @@
...
+
+      /**
+       * Multiply the vector with a scalar
+       */
+      Point<dim> operator*(const double scalar) const;
+
+      /**
+       * add two points
+       */
+      Point<dim> operator+(const Point<dim> &point) const;
```
Comment (tjhei) [inline contrib/world_builder/include/world_builder/point.h]: binary operators are typically implemented using friends and based on += etc. See https://en.cppreference.com/w/cpp/language/operators

File: contrib/world_builder/include/world_builder/point.h [hunk 15]
```diff
@@ -0,0 +1,167 @@
...
+
+      /**
+       * access index
+       */
+      double &operator[](const unsigned int index);
+
+      /**
+       * access index (const)
+       */
+      const double &operator()(const unsigned int index) const;
```
Comment (tjhei) [inline contrib/world_builder/include/world_builder/point.h]: in deal.II we are sloppy and allow [] and (). Maybe you don't want to go there here?

File: contrib/world_builder/include/world_builder/point.h [hunk 16]
```diff
@@ -0,0 +1,167 @@
...
+      /**
+       * access index
+       */
+      double &operator()(const unsigned int index);
+
+
+      /**
+       * return the internal array which stores the point data.
+       */
+      std::array<double,dim> get_array() const;
```
Comment (tjhei) [inline contrib/world_builder/include/world_builder/point.h]: const &?

File: contrib/world_builder/include/world_builder/point.h [hunk 17]
```diff
@@ -0,0 +1,167 @@
...
+
+
+      /**
+       * returns the coordinate system associated with the data.
+       */
+      CoordinateSystem get_coordinate_system() const;
```
Comment (tjhei) [inline contrib/world_builder/include/world_builder/point.h]: const &?

File: contrib/world_builder/include/world_builder/utilities.h [hunk 18]
```diff
@@ -0,0 +1,181 @@
...
+         * Constructor based on providing the geometry model as a pointer
+         */
+        NaturalCoordinate(const Point<3> &position,
+                          const ::WorldBuilder::CoordinateSystems::Interface &coordinate_system);
+
+        /**
+         * Returns the coordinates in the given coordinate system, which may
+         * not be Cartesian.
+         */
+        std::array<double,3> &get_coordinates();
```
Comment (tjhei) [inline contrib/world_builder/include/world_builder/utilities.h]: const & return value and make the method const?

File: contrib/world_builder/include/world_builder/utilities.h [hunk 19]
```diff
@@ -0,0 +1,181 @@
...
+         * The coordinate that represents the 'surface' directions in the
+         * chosen coordinate system.
+         */
+        std::array<double,2> get_surface_coordinates() const;
+
+        /**
+         * The coordinate that represents the 'depth' direction in the chosen
+         * coordinate system.
+         */
+        double get_depth_coordinate() const;
```
Comment (tjhei) [inline contrib/world_builder/include/world_builder/utilities.h]: value or coordinate?

File: contrib/world_builder/include/world_builder/utilities.h [hunk 20]
```diff
@@ -0,0 +1,181 @@
...
+     */
+    CoordinateSystem
+    string_to_coordinate_system (const std::string &);
+
+
+    /**
+     * Convert point to array
+     */
+    template<int dim>
+    std::array<double,dim> convert_point_to_array(Point<dim>);
```
Comment (tjhei) [inline contrib/world_builder/include/world_builder/utilities.h]: ``const Point<dim> &point``

File: contrib/world_builder/include/world_builder/utilities.h [hunk 21]
```diff
@@ -0,0 +1,181 @@
...
+
+    /**
+     * Converts a string to a double
+     */
+    double
+    string_to_int(const std::string &string);
+
+
+    /**
+     * Converts a string to a double
```
Comment (tjhei) [inline contrib/world_builder/include/world_builder/utilities.h]: copy&paste error

File: contrib/world_builder/include/world_builder/utilities.h [hunk 22]
```diff
@@ -0,0 +1,181 @@
...
+    std::array<double,dim> convert_point_to_array(Point<dim>);
+
+    /**
+     * Converts a string to a double
+     */
+    double
+    string_to_double(const std::string &string);
+
+    /**
+     * Converts a string to a double
```
Comment (tjhei) [inline contrib/world_builder/include/world_builder/utilities.h]: copy and paste error

File: contrib/world_builder/include/world_builder/utilities.h [hunk 23]
```diff
@@ -0,0 +1,181 @@
...
+     * Convert point to array
+     */
+    template<int dim>
+    std::array<double,dim> convert_point_to_array(Point<dim>);
+
+    /**
+     * Converts a string to a double
+     */
+    double
+    string_to_double(const std::string &string);
```
Comment (tjhei) [inline contrib/world_builder/include/world_builder/utilities.h]: how about a ``template <T> to_string(const T&)`` function instead? See deal.II.

File: contrib/world_builder/include/world_builder/world.h [hunk 24]
```diff
@@ -0,0 +1,102 @@
...
+       * Destructor
+       */
+      ~World();
+
+      /**
+       * read in the world builder file
+       */
+      void read(ptree &property_tree);
+
+      double temperature(const std::array<double, 2> point, const double depth, const double gravity_norm) const;
```
Comment (tjhei) [inline contrib/world_builder/include/world_builder/world.h]: pass arrary as const&

File: contrib/world_builder/include/world_builder/world.h [hunk 25]
```diff
@@ -0,0 +1,102 @@
...
+
+      double temperature(const std::array<double, 3> point, const double depth, const double gravity_norm) const;
```
Comment (tjhei) [inline contrib/world_builder/include/world_builder/world.h]: what is the difference of the 2 and 3 version?

File: contrib/world_builder/source/coordinate_system/cartesian.cc [hunk 26]
```diff
@@ -0,0 +1,118 @@
...
+#include <world_builder/coordinate_systems/cartesian.h>
+
+namespace WorldBuilder
+{
+  namespace CoordinateSystems
+  {
+    Cartesian::Cartesian()
+    {}
+
+    // todo: add relative path somehow, to output when there are errors
```
Comment (tjhei) [inline contrib/world_builder/source/coordinate_system/cartesian.cc]: fix

File: contrib/world_builder/source/point.cc [hunk 27]
```diff
@@ -0,0 +1,262 @@
...
+
+#include <world_builder/point.h>
+#include <world_builder/assert.h>
+
+namespace WorldBuilder
+{
+  template<>
+  Point<3>::Point()
+    :
+    point({0,0,0})
```
Comment (tjhei) [inline contrib/world_builder/source/point.cc]: coordinate system=?

File: contrib/world_builder/source/world.cc [hunk 28]
```diff
@@ -0,0 +1,153 @@
...
+#include <world_builder/assert.h>
+#include <world_builder/point.h>
+
+
+namespace WorldBuilder
+{
+  World::World(std::string filename)
+    :
+    potential_mantle_temperature(1600),
+    thermal_expansion_coefficient_alfa(3.5e-5),
```
Comment (tjhei) [inline contrib/world_builder/source/world.cc]: alpha?

File: doc/indent [hunk 29]
```diff
@@ -46,10 +46,10 @@ fi
 # with up to 10 in parallel
 echo "--- Indenting all ASPECT header and source files"
 
-find tests include source benchmarks cookbooks unit_tests \( -name '*.cc' -o -name '*.h' \) -print | xargs -n 50 -P 10 astyle --options=doc/astyle.rc
+find tests include source benchmarks cookbooks unit_tests contrib/world_builder/include contrib/world_builder/source \( -name '*.cc' -o -name '*.h' \) -print | xargs -n 50 -P 10 astyle --options=doc/astyle.rc
```
Comment (tjhei) [inline doc/indent]: it is enough to specify ``contrib/world_builder`` once

File: doc/indent [hunk 30]
```diff
@@ -46,10 +46,10 @@ fi
...
 
 # remove execute permission on source files:
-find tests include source benchmarks cookbooks unit_tests \( -name '*.cc' -o -name '*.h' -o -name '*.prm' \) -print | xargs -n 50 -P 10 chmod -x
+find tests include source benchmarks cookbooks unit_tests contrib/world_builder/include contrib/world_builder/source \( -name '*.cc' -o -name '*.h' -o -name '*.prm' \) -print | xargs -n 50 -P 10 chmod -x
```
Comment (tjhei) [inline doc/indent]: same

File: doc/indent [hunk 31]
```diff
@@ -61,4 +61,4 @@ dos_to_unix()
     rm -f $f.tmp
 }
 export -f dos_to_unix
-find tests include source unit_tests \( -name '*.cc' -o -name '*.h' -o -name '*.prm' \) -print | xargs -n 1 -P 10 -I {} bash -c 'dos_to_unix "$@"' _ {} 
+find tests include source unit_tests contrib/world_builder/include contrib/world_builder/source \( -name '*.cc' -o -name '*.h' -o -name '*.prm' \) -print | xargs -n 1 -P 10 -I {} bash -c 'dos_to_unix "$@"' _ {}
```
Comment (tjhei) [inline doc/indent]: same

File: include/aspect/initial_temperature/world_builder.h [hunk 32]
```diff
@@ -0,0 +1,84 @@
...
+
+namespace aspect
+{
+  namespace InitialTemperature
+  {
+    using namespace dealii;
+
+    /**
+     * A class that implements temperature initial conditions based on a
+     * functional description provided in the input file.
```
Comment (tjhei) [inline include/aspect/initial_temperature/world_builder.h]: reference worldbuilder?

File: source/initial_composition/world_builder.cc [hunk 33]
```diff
@@ -0,0 +1,68 @@
...
+// explicit instantiations
+namespace aspect
+{
+  namespace InitialComposition
+  {
+    ASPECT_REGISTER_INITIAL_COMPOSITION_MODEL(WorldBuilder,
+                                              "world builder",
+                                              "Specify the composition in terms of an explicit formula. The format of these "
+                                              "functions follows the syntax understood by the "
+                                              "muparser library, see Section~\\ref{sec:muparser-format}.")
```
Comment (tjhei) [inline source/initial_composition/world_builder.cc]: update!

File: source/initial_temperature/world_builder.cc [hunk 34]
```diff
@@ -0,0 +1,72 @@
...
+namespace aspect
+{
+  namespace InitialTemperature
+  {
+    ASPECT_REGISTER_INITIAL_TEMPERATURE_MODEL(WorldBuilder,
+                                              "world builder",
+                                              "Specify the initial temperature in terms of an "
+                                              "explicit formula. The format of these "
+                                              "functions follows the syntax understood by the "
+                                              "muparser library, see Section~\\ref{sec:muparser-format}.")
```
Comment (tjhei) [inline source/initial_temperature/world_builder.cc]: update!

File: unit_tests/world_builder_point.cc [hunk 35]
```diff
@@ -0,0 +1,60 @@
...
+  You should have received a copy of the GNU General Public License
+  along with ASPECT; see the file LICENSE.  If not see
+  <http://www.gnu.org/licenses/>.
+*/
+
+#include "common.h"
+#include <world_builder/point.h>
+
+using namespace WorldBuilder;
+using Catch::Matchers::Contains;
```
Comment (tjhei) [inline unit_tests/world_builder_point.cc]: you should be able to remove this now.

File: contrib/world_builder/include/world_builder/point.h [hunk 36]
```diff
@@ -0,0 +1,167 @@
...
+      /**
+      * Computes the square of the norm, which is the sum of the absolute squares
+      * x_i * x_i + y_i * y_i + z_i * z_i in 3d.
+      */
+      double norm_square() const;
+
+
+    private:
+      std::array<double,dim> point;
+      CoordinateSystem coordinate_system;
```
Comment (tjhei) [inline contrib/world_builder/include/world_builder/point.h]: you point class also contains a coordinate system. Maybe a name like "location" or something would be more appropriate?

**Comment by bangerth** [comment]
Let me know when you're done with revisions for @tjhei 's comments and I can take a look as well!

**Comment by MFraters** [comment]
@tjhei Thanks for the quick reviewing!
@bangerth Sure :)

File: contrib/world_builder/include/world_builder/features/continental_plate.h (hunk 6 above)
Comment (MFraters) [reply contrib/world_builder/include/world_builder/features/continental_plate.h]: Yes, that makes a lot of sence

File: contrib/world_builder/include/world_builder/point.h (hunk 15 above)
Comment (MFraters) [reply contrib/world_builder/include/world_builder/point.h]: do you have a preference for either?

File: contrib/world_builder/include/world_builder/point.h (hunk 36 above)
Comment (MFraters) [reply contrib/world_builder/include/world_builder/point.h]: I do not fully understand what you mean. Do you want to rename `CoordinateSystem coordinate_system;` to `CoordinateSystem location;`? That name doesn't make much sense to me.

File: contrib/world_builder/include/world_builder/point.h (hunk 13 above)
Comment (MFraters) [reply contrib/world_builder/include/world_builder/point.h]: That doesn't seem to work, are you sure? Let's talk when you are done with your bike ;)

File: contrib/world_builder/include/world_builder/point.h (hunk 17 above)
Comment (MFraters) [reply contrib/world_builder/include/world_builder/point.h]: It is just an enum, I don't think that is necessary.

File: contrib/world_builder/include/world_builder/utilities.h (hunk 19 above)
Comment (MFraters) [reply contrib/world_builder/include/world_builder/utilities.h]: it is the depth value of the coordinate, which is to make it more complicated defined from the bottom/center outwards. Rename?

File: contrib/world_builder/include/world_builder/utilities.h (hunk 23 above)
Comment (MFraters) [reply contrib/world_builder/include/world_builder/utilities.h]: hmm, that is interesting indeed, but the name you have written down is the wrong way around. It should be something like `from_string` or `string_to`. What do you think?

File: contrib/world_builder/include/world_builder/world.h (hunk 25 above)
Comment (MFraters) [reply contrib/world_builder/include/world_builder/world.h]: The 2d version converts the 2d point to a 3d point and calls the 3d function.

**Comment by MFraters** [comment]
@tjhei I think I have addressed all your comments, but there are still some questions open.

@bangerth You can review it now :)

**Comment by MFraters** [comment]
I have added 2d (only a few lines of code) with a new test and cleanup up some of the reading functionality with a separate function.

**Comment by MFraters** [comment]
ah, I forgot to add figures for the 2d cross section. First figure, the 3d model, where the green line is where the cross section is taken:
screenshot from 2018-07-03 18-39-23

Secondly, the cross section itself:
screenshot from 2018-07-03 18-41-29

**Comment by MFraters** [comment]
superseded by #2571

**Pull request closed without merging**